🐛 Fix ClusterCacheTracker memory leak

ionutbalutoiu opened the pull request:
Decouple the code that creates uncached client from the code that creates the cached client. This way, we don't start the cache until the cached client is created.

This avoids scenarios where the `ClusterCacheTracker` cache is started when only an uncached client is needed.

There is an existing code path, described in the original issue, where the `ClusterCacheTracker` cache is started when only an uncached client is needed. The cache is re-created later, and the initial cache is still running continuously in the background.

<!-- Thanks for sending a pull request! Here are some tips for you:
    1. If this is your first time, please read our contributor guidelines: https://github.com/kubernetes-sigs/cluster-api/blob/main/CONTRIBUTING.md#contributing-a-patch and developer guide https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/book/src/developer/guide.md

    2. Please add an icon to the title of this PR (see https://sigs.k8s.io/cluster-api/CONTRIBUTING.md#contributing-a-patch), and delete this line and similar ones
    the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) 
-->

**What this PR does / why we need it**:

Fix the memory leak described in the issue https://github.com/kubernetes-sigs/cluster-api/issues/9542, when cluster `kubeconfig` is rotated frequently,

**Which issue(s) this PR fixes**:
Fixes #9542

<!-- 
Please label this pull request according to what area(s) you are addressing. For reference on PR/issue labels, see: https://github.com/kubernetes-sigs/cluster-api/labels?q=area+

Area example:
/area runtime-sdk
-->

killianmuldoon reviewed (commented):
/area clustercachetracker

/ok-to-test

ionutbalutoiu commented:
Fixed the unused context `golangci-lint` errors via the following changes:
https://github.com/kubernetes-sigs/cluster-api/compare/36c0b2c6d800d04cfc1e299fa24da540756d05a8..88e53de8cabeac51697ee29c297a427bfdd361ad

killianmuldoon commented:
/test

killianmuldoon commented:
/test pull-cluster-api-e2e-main

killianmuldoon commented:
/test pull-cluster-api-e2e-workload-upgrade-1-28-latest-main

vincepri commented:
/assign

Would like some time to review this PR in a bit more details

vincepri commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -374,26 +384,58 @@ func (t *ClusterCacheTracker) runningOnWorkloadCluster(ctx context.Context, c cl
 
-// createClient creates a cached client, and uncached client and a mapper based on a rest.Config.
-func (t *ClusterCacheTracker) createClient(ctx context.Context, config *rest.Config, cluster client.ObjectKey, indexes []Index) (client.Client, client.Client, *stoppableCache, error) {
+// createHTTPClientAndMapper creates a http client and a dynamic rest mapper for the given cluster, based on the rest.Config.
+func (t *ClusterCacheTracker) createHTTPClientAndMapper(config *rest.Config, cluster client.ObjectKey) (*http.Client, meta.RESTMapper, error) {
```
Given that now we have this function, could we create an http client and rest mapper for each config only one and eventually use it also for the cache? Basically, `createUncachedClient` and `createCachedClient` should get those two fields as input, instead of the *rest.Config

ionutbalutoiu replied:
Well, I believe we still need the `*rest.Config` parameter.

For the `createUncachedClient` function, we need it for `client.New` call.

And, for the `createCachedClient` function, we need it for:
* `cache.New` call.
* `client.New` call.
* `newClientWithTimeout` (since it uses uses `config.Timeout` when wrapping the client before returning it).
* `go t.healthCheckCluster` goroutine.

However, it makes sense moving the `createHTTPClientAndMapper` call outside `createUncachedClient` / `createCachedClient` and adding the `httpClient` and `mapper` as parameters to the get client functions.

ionutbalutoiu replied:
Actually, I don't think that it solves anything moving the `createHTTPClientAndMapper` outside the get client functions.

We would still have to call this twice anyway:
* Before the `createUncachedClient`.
* Before `createCachedClient`. This second call is needed since the config is changed when `in-cluster` config is used, and re-using the `httpClient` and `mapper` from the `createUncachedClient` call is not what we want. 

I would let it as it is, since we cannot remove the `*rest.Config` parameter (as pointed above).

What do you think?

vincepri replied:
I'm confused. Why can't we share the `mapper` and `httpClient` across both calls? If anything, we shouldn't duplicate these inputs for the two clients for the same cluster

ionutbalutoiu replied:
> I'm confused. Why can't we share the `mapper` and `httpClient` across both calls? If anything, we shouldn't duplicate these inputs for the two clients for the same cluster

We can share them, but it will only work only when `runningOnCluster` is `false`. The `uncachedClient` decides if the cached client with in-cluster config is needed.

If `runningOnCluster` is `true`, we need to create a different `mapper` and `httpClient` using the in-cluster config, and the previous `mapper` and `httpClient` cannot be re-used. I can update the code as suggested, but in the if branch with `runningOnCluster == true` I need override the `mapper` and `httpClient` using the in-cluster config.

Besides this, as mentioned above, we cannot remove `*rest.Config` parameter from `createCachedClient` function as suggested initially (it's used elsewhere besides calling `createHTTPClientAndMapper`). So, what I would do is adding two more extra parameters to the `createCachedClient` fuction, without removing any existing ones.

If you still believe it's better to re-use the `mapper` and `httpClient` from uncached client (but override them in the `runningOnCluster == true` if branch, I will make the change.

ionutbalutoiu replied:
See this piece of code for reference:
https://github.com/ionutbalutoiu/cluster-api/blob/8cf7a6632195f42d43da6eebb693286444e28f70/controllers/remote/cluster_cache_tracker.go#L294-L337

ionutbalutoiu commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -321,13 +325,19 @@ func (t *ClusterCacheTracker) newClusterAccessor(ctx context.Context, cluster cl
 
-		// Create a new client and overwrite the previously created client.
-		c, _, cache, err = t.createClient(ctx, config, cluster, indexes)
+		// Create a client and a cache for the cluster, using the in-cluster config.
+		c, cache, err = t.createCachedClient(ctx, config, cluster, indexes)
 		if err != nil {
 			return nil, errors.Wrap(err, "error creating client for self-hosted cluster")
 		}
```
I agree. I will move these outside to remove the duplicated code.

vincepri commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -321,16 +325,17 @@ func (t *ClusterCacheTracker) newClusterAccessor(ctx context.Context, cluster cl
 	}
 
+	// Create a client and a cache for the cluster.
+	c, cache, err = t.createCachedClient(ctx, config, cluster, indexes)
```
```suggestion
	c, cache, err := t.createCachedClient(ctx, config, cluster, indexes)
```

And remove the variable declaration above?

ionutbalutoiu replied:
Agree!

vincepri commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -374,26 +379,58 @@ func (t *ClusterCacheTracker) runningOnWorkloadCluster(ctx context.Context, c cl
+}
+
+// createCachedClient creates a cached client for the given cluster, based on a rest.Config.
+func (t *ClusterCacheTracker) createCachedClient(ctx context.Context, config *rest.Config, cluster client.ObjectKey, indexes []Index) (client.Client, *stoppableCache, error) {
```
Avoid returning two parameters + errors, wrap them in an output struct

vincepri replied:
Example:

```golang
type cachedClientOutput struct {
  Client client.Client
  Cache *stoppableCache
}
```

ionutbalutoiu replied:
To stay consistent to the current codebase, I just re-used the same function signature that we have currently in `main` branch:
https://github.com/kubernetes-sigs/cluster-api/blob/83d79695ed028bd6ae6f1aaa1b5cd19eaa2ac267/controllers/remote/cluster_cache_tracker.go#L377-L378

but removing the uncached client from return values, and renaming it accordingly.

I will make this change as well.

ionutbalutoiu commented:
@vincepri Implemented the following changes for your most recent code review:
https://github.com/kubernetes-sigs/cluster-api/compare/8cf7a6632195f42d43da6eebb693286444e28f70..500968845d650521109759790ce79785879b70eb

Thank-you!

sbueringer commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -374,26 +376,63 @@ func (t *ClusterCacheTracker) runningOnWorkloadCluster(ctx context.Context, c cl
+	// Create a http client and a mapper for the cluster.
+	httpClient, mapper, err := t.createHTTPClientAndMapper(config, cluster)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error creating client for remote cluster %q", cluster.String())
```
```suggestion
		return nil, errors.Wrapf(err, "error creating cached client for remote cluster %q", cluster.String())
```

sbueringer commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -417,7 +456,7 @@ func (t *ClusterCacheTracker) createClient(ctx context.Context, config *rest.Con
 	for _, index := range indexes {
 		if err := cache.IndexField(ctx, index.Object, index.Field, index.ExtractValue); err != nil {
-			return nil, nil, nil, errors.Wrapf(err, "error adding index for field %q to cache for remote cluster %q", index.Field, cluster.String())
+			return nil, errors.Wrapf(err, "error adding index for field %q to cache for remote cluster %q", index.Field, cluster.String())
```
```suggestion
			return nil, errors.Wrapf(err, "error creating cached client for remote cluster %q: error adding index for field %q to cache", cluster.String(), index.Field)
```

sbueringer reviewed (commented):
Some smaller findings and answered on the discussion with some more context

sbueringer commented:
@ionutbalutoiu Sorry for the long delay. I sort of lost track of this PR a bit. Happy to quickly review again and merge once the open points are addressed

ionutbalutoiu commented on controllers/remote/cluster_cache_tracker.go:
```diff
@@ -374,26 +376,63 @@ func (t *ClusterCacheTracker) runningOnWorkloadCluster(ctx context.Context, c cl
+	return uncachedClient, nil
+}
+
+type cachedClientOutput struct {
```
The output struct change was suggested by @vincepri in [this previous PR comment](https://github.com/kubernetes-sigs/cluster-api/pull/9543#discussion_r1395979323).

I'm fine with anything as long as it ready to be merged.

sbueringer replied:
All good :)

ionutbalutoiu commented:
> @ionutbalutoiu Sorry for the long delay. I sort of lost track of this PR a bit. Happy to quickly review again and merge once the open points are addressed

Thank-you for taking the time to review this! I've addressed all the remaining review open items. I also did a rebase against latest `main` branch.

Please take a look at the latest PR changes and let me know if there's anything else.

sbueringer commented:
Thank you very much!

Let's see if we have some luck with cherry-picking

/lgtm
/approve

sbueringer commented:
/cherry-pick release-1.6
